File: node/packages/botas-core/src/posthog-telemetry.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { createHash } from 'node:crypto'
import { VERSION } from './version.js'

/**
 * PostHog client interface (subset used by botas).
 * @internal
 */
interface PostHogClient {
  capture: (options: {
    distinctId: string
    event: string
    properties?: Record<string, unknown>
  }) => void
  shutdown: () => Promise<void>
}

/**
 * PostHog module interface (subset used by botas).
 * @internal
 */
interface PostHogModule {
  PostHog: new (
    apiKey: string,
    options?: {
      host?: string
      disableGeoip?: boolean
      flushAt?: number
      flushInterval?: number
    }
  ) => PostHogClient
}

// Optional peer dependency. Loaded via dynamic import so the module works in
// environments without `posthog-node` (and avoids `createRequire`,
// which fails on Deno when modules are loaded from a remote URL like JSR).
let _posthogModule: PostHogModule | null = null
let _posthogClient: PostHogClient | null = null
let _isDisabled = false // true = env var not set or module unavailable
let _isInitialized = false
let _distinctId: string | null = null
let _botStartedEmitted = false

/** @internal Reset telemetry state — for testing only. */
export function _resetPostHogTelemetry(): void {
  _posthogModule = null
  _posthogClient = null
  _isDisabled = false
  _isInitialized = false
  _distinctId = null
  _botStartedEmitted = false
}

/**
 * Lazy-initialize PostHog client on first trackEvent call.
 * @internal
 */
function initializePostHog(): void {
  if (_isInitialized) return
  _isInitialized = true

  // Read env vars
  const apiKey = process.env.POSTHOG_API_KEY
  const host = process.env.POSTHOG_HOST ?? 'https://us.i.posthog.com'

  // If no API key, disable telemetry (zero-cost no-op)
  if (!apiKey || apiKey.trim() === '') {
    _isDisabled = true
    return
  }

  // Try to load posthog-node module (dynamic import, type-erased at runtime)
  // @ts-expect-error - posthog-node is an optional peer dependency; types may not be available
  import('posthog-node')
    .then((module: unknown) => {
      _posthogModule = module as PostHogModule
      // Create PostHog client
      _posthogClient = new _posthogModule.PostHog(apiKey, {
        host,
        disableGeoip: true, // no IP-based location
        flushInterval: 30000, // 30s batch interval
        flushAt: 20, // or every 20 events
      })

      // Compute distinct_id from CLIENT_ID
      const clientId = process.env.CLIENT_ID
      if (clientId && clientId.trim() !== '') {
        const hash = createHash('sha256').update(clientId).digest('hex')
        _distinctId = hash.substring(0, 16)
      } else {
        _distinctId = 'botas-anonymous'
      }
    })
    .catch(() => {
      // posthog-node not available — disable telemetry
      _isDisabled = true
    })
}

/**
 * Get runtime version string for Node.js.
 * @internal
 */
function getRuntimeVersion(): string {
  return `Node ${process.version}`
}

/**
 * Get sanitized channel type hint.
 * @internal
 */
function getChannelType(channelId: string | undefined): string {
  if (!channelId) return 'other'
  const lower = channelId.toLowerCase()
  if (lower === 'emulator') return 'emulator'
  if (lower === 'msteams') return 'msteams'
  if (lower === 'webchat') return 'webchat'
  return 'other'
}

/**
 * Track a PostHog event (fire-and-forget).
 * @internal
 */
export function trackEvent(name: string, properties: Record<string, unknown>): void {
  // Lazy-init on first call
  if (!_isInitialized) {
    initializePostHog()
  }

  // If disabled or client not ready, no-op
  if (_isDisabled || !_posthogClient || !_distinctId) {
    return
  }

  // Add common properties
  const allProperties = {
    sdk_language: 'node',
    sdk_version: VERSION,
    runtime_version: getRuntimeVersion(),
    ...properties,
  }

  // Fire-and-forget (never block pipeline)
  try {
    _posthogClient.capture({
      distinctId: _distinctId,
      event: name,
      properties: allProperties,
    })
  } catch {
    // Swallow all PostHog errors silently
  }
}

/**
 * Track `botas/bot_started` event (once per process lifetime).
 * @internal
 */
export function trackBotStarted(options: {
  handlerCount: number
  invokeHandlerCount: number
  middlewareCount: number
  hasCatchAll: boolean
  hasStateStorage: boolean
  authFlow: 'client_credentials' | 'managed_identity' | 'none'
}): void {
  if (_botStartedEmitted) return
  _botStartedEmitted = true

  trackEvent('botas/bot_started', {
    handler_count: options.handlerCount,
    invoke_handler_count: options.invokeHandlerCount,
    middleware_count: options.middlewareCount,
    has_catch_all: options.hasCatchAll,
    has_state_storage: options.hasStateStorage,
    auth_flow: options.authFlow,
  })
}

/**
 * Track `botas/activity_received` event.
 * @internal
 */
export function trackActivityReceived(options: {
  activityType: string
  hasHandler: boolean
  channelId: string | undefined
}): void {
  trackEvent('botas/activity_received', {
    activity_type: options.activityType,
    has_handler: options.hasHandler,
    channel_type: getChannelType(options.channelId),
  })
}

/**
 * Track `botas/handler_dispatched` event.
 * @internal
 */
export function trackHandlerDispatched(options: {
  activityType: string
  dispatchMode: 'type' | 'invoke' | 'catchall'
  durationMs: number
}): void {
  trackEvent('botas/handler_dispatched', {
    activity_type: options.activityType,
    dispatch_mode: options.dispatchMode,
    duration_ms: options.durationMs,
  })
}

/**
 * Track `botas/handler_error` event.
 * @internal
 */
export function trackHandlerError(options: {
  activityType: string
  errorType: string
}): void {
  trackEvent('botas/handler_error', {
    activity_type: options.activityType,
    error_type: options.errorType,
  })
}

/**
 * Track `botas/outbound_sent` event.
 * @internal
 */
export function trackOutboundSent(options: {
  operation: 'send' | 'update' | 'delete' | 'create_conversation'
  success: boolean
}): void {
  trackEvent('botas/outbound_sent', {
    operation: options.operation,
    success: options.success,
  })
}

/**
 * Flush buffered events on process exit (best-effort).
 * @internal
 */
export function flushOnShutdown(): void {
  if (_posthogClient) {
    _posthogClient.shutdown().catch(() => {
      // Ignore flush errors
    })
  }
}

// Register shutdown handler
process.on('exit', () => {
  flushOnShutdown()
})

File: python/packages/botas/src/botas/_posthog_telemetry.py
"""PostHog usage telemetry for botas.

Anonymous, opt-in telemetry to understand SDK adoption and feature usage.
Completely disabled when POSTHOG_API_KEY is unset. No PII, no message text,
fire-and-forget.
"""

from __future__ import annotations

import atexit
import hashlib
import os
import sys
import threading
from typing import TYPE_CHECKING, Any, Optional

if TYPE_CHECKING:
    from posthog import Posthog

_client: Optional[Posthog] = None
_is_initialized: bool = False
_is_disabled: bool = False
_lock = threading.Lock()
_distinct_id: Optional[str] = None
_bot_started_emitted: bool = False


def _get_distinct_id() -> str:
    """Derive distinct_id from CLIENT_ID env var using SHA-256 (first 16 hex chars)."""
    client_id = os.environ.get("CLIENT_ID", "").strip()
    if not client_id:
        return "botas-anonymous"
    return hashlib.sha256(client_id.encode("utf-8")).hexdigest()[:16]


def _get_common_properties() -> dict[str, Any]:
    """Return common event properties: sdk_language, sdk_version, runtime_version."""
    from botas._version import __version__

    python_version = f"Python {sys.version_info.major}.{sys.version_info.minor}"
    return {
        "sdk_language": "python",
        "sdk_version": __version__,
        "runtime_version": python_version,
    }


def _get_channel_type(channel_id: Optional[str]) -> str:
    """Sanitize channel_id into a known type: emulator, msteams, webchat, or other."""
    if not channel_id:
        return "other"
    lower = channel_id.lower()
    if lower == "emulator":
        return "emulator"
    if lower == "msteams":
        return "msteams"
    if lower == "webchat":
        return "webchat"
    return "other"


def _initialize_client() -> None:
    """Initialize PostHog client once, or set disabled flag if unavailable."""
    global _client, _is_initialized, _is_disabled, _distinct_id
    if _is_initialized:
        return

    _is_initialized = True
    api_key = os.environ.get("POSTHOG_API_KEY", "").strip()
    if not api_key:
        _is_disabled = True
        return

    try:
        from posthog import Posthog

        host = os.environ.get("POSTHOG_HOST", "https://eu.i.posthog.com").strip()
        _client = Posthog(
            api_key,
            host=host,
            # No IP-based geolocation
            disable_geoip=True,
            # Batch events, don't block
            flush_interval=30.0,
            flush_at=20,
        )
        # Compute distinct_id from CLIENT_ID env var
        _distinct_id = _get_distinct_id()
        # Register shutdown hook for best-effort flush
        atexit.register(_shutdown)
    except ImportError:
        _is_disabled = True
        _client = None


def _shutdown() -> None:
    """Flush buffered events on process exit."""
    global _client
    if _client is not None:
        try:
            _client.flush()
            _client.shutdown()
        except Exception:
            pass


def track_event(event_name: str, properties: dict[str, Any]) -> None:
    """Track a PostHog event with common properties. Fire-and-forget, never throws.

    Args:
        event_name: Event name (e.g., "botas/bot_started").
        properties: Event-specific properties.
    """
    global _is_disabled, _client, _distinct_id
    if _is_disabled:
        return

    with _lock:
        _initialize_client()
        if _is_disabled or _client is None or _distinct_id is None:
            return

    try:
        # Merge common properties
        merged = {**_get_common_properties(), **properties}
        _client.capture(_distinct_id, event_name, merged)
    except Exception:
        # Silently swallow telemetry errors
        pass


def track_bot_started(
    handler_count: int,
    invoke_handler_count: int,
    middleware_count: int,
    has_catch_all: bool,
    has_state_storage: bool,
    auth_flow: str,
) -> None:
    """Emit botas/bot_started event once per process lifetime."""
    global _bot_started_emitted
    if _bot_started_emitted:
        return
    _bot_started_emitted = True
    track_event(
        "botas/bot_started",
        {
            "handler_count": handler_count,
            "invoke_handler_count": invoke_handler_count,
            "middleware_count": middleware_count,
            "has_catch_all": has_catch_all,
            "has_state_storage": has_state_storage,
            "auth_flow": auth_flow,
        },
    )


def track_activity_received(
    activity_type: str,
    has_handler: bool,
    channel_id: Optional[str],
) -> None:
    """Emit botas/activity_received event once per turn."""
    track_event(
        "botas/activity_received",
        {
            "activity_type": activity_type,
            "has_handler": has_handler,
            "channel_type": _get_channel_type(channel_id),
        },
    )


def track_handler_dispatched(
    activity_type: str,
    dispatch_mode: str,
    duration_ms: int,
) -> None:
    """Emit botas/handler_dispatched event when a handler executes."""
    track_event(
        "botas/handler_dispatched",
        {
            "activity_type": activity_type,
            "dispatch_mode": dispatch_mode,
            "duration_ms": duration_ms,
        },
    )


def track_handler_error(
    activity_type: str,
    error_type: str,
) -> None:
    """Emit botas/handler_error event when a handler throws."""
    track_event(
        "botas/handler_error",
        {
            "activity_type": activity_type,
            "error_type": error_type,
        },
    )


def track_outbound_sent(
    operation: str,
    success: bool,
) -> None:
    """Emit botas/outbound_sent event when an outbound API call completes."""
    track_event(
        "botas/outbound_sent",
        {
            "operation": operation,
            "success": success,
        },
    )
